semantic/check_test.go

```diff
@@ -0,0 +1,1161 @@
+								},
+							},
+							&OptionStatement{
+								Assignment: &NativeVariableAssignment{
+									Identifier: &Identifier{Name: "a"},
+									Init:       &IntegerLiteral{Value: 1},
+								},
+							},
+							&OptionStatement{
+								Assignment: &NativeVariableAssignment{
+									Identifier: &Identifier{Name: "b"},
+									Init:       &IntegerLiteral{Value: 2},
+								},
+							},
+							&NativeVariableAssignment{
+								Identifier: &Identifier{Name: "b"},
+								Init:       &IntegerLiteral{Value: 3},
+							},
+						},
+					},
+				},
+			},
+			err: fmt.Errorf(`var "b" redeclared at 0:0-0:0`),
+		},
+		{
+			// package foo
+			// a = 0
+			// a = 1
+			//
+			name: "redeclaration",
+			pkg: &Package{
+				Package: "foo",
+				Files: []*File{
+					{
+						Package: &PackageClause{
+							Name: &Identifier{Name: "foo"},
+						},
+						Body: []Statement{
+							&NativeVariableAssignment{
+								Identifier: &Identifier{Name: "a"},
+								Init:       &IntegerLiteral{Value: 0},
+							},
+							&NativeVariableAssignment{
+								Identifier: &Identifier{Name: "a"},
+								Init:       &IntegerLiteral{Value: 1},
+							},
+						},
+					},
+				},
+			},
+			err: fmt.Errorf(`var "a" redeclared at 0:0-0:0`),
+		},
+		{
+			// package foo
+			// a = 0
+			// f = () => {
+			//   a = 1
+			//   b = a
+			//   b = 1
+			//   return b
+			// }
+			//
+			name: "redeclaration inside function",
+			pkg: &Package{
+				Package: "foo",
+				Files: []*File{
+					{
+						Package: &PackageClause{
+							Name: &Identifier{Name: "foo"},
+						},
+						Body: []Statement{
+							&NativeVariableAssignment{
+								Identifier: &Identifier{Name: "a"},
+								Init:       &IntegerLiteral{Value: 0},
+							},
+							&NativeVariableAssignment{
+								Identifier: &Identifier{Name: "f"},
+								Init: &FunctionExpression{
+									Block: &FunctionBlock{
+										Body: &Block{
+											Body: []Statement{
+												&NativeVariableAssignment{
+													Identifier: &Identifier{Name: "a"},
+													Init:       &IntegerLiteral{Value: 1},
+												},
+												&NativeVariableAssignment{
+													Identifier: &Identifier{Name: "b"},
+													Init:       &IdentifierExpression{Name: "a"},
+												},
+												&NativeVariableAssignment{
+													Identifier: &Identifier{Name: "b"},
+													Init:       &IntegerLiteral{Value: 1},
+												},
+												&ReturnStatement{
+													Argument: &IdentifierExpression{Name: "b"},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			err: fmt.Errorf(`var "b" redeclared at 0:0-0:0`),
+		},
+		{
+			// package foo
+			// f = (a) => {
+			//   a = 1
+			//   return a
+			// }
+			//
+			name: "reassign parameter",
+			skip: "reassigning a function parameter inside a function block is a variable redeclaration and should error",
```

Comment:
Why is this one skipped?

Reply:
I was just creating an issue for this. Currently the function body is in a separate block nested inside of the block where the function parameters are assigned. This means you can reassign (in this case it would be shadowing) a function's parameter inside its body. Technically this is not correct; the parameters should be declared in the same block as the body and so should not be able to be reassigned.

Reply:
OK, so the fix will likely be to update the ScopedVisitor to not create a new scope for parameters vs the body. Sounds good, makes sense as a separate issue.